### DISCUSSION TITLE
Why do lost targets remain visible in the Cryostat target list?

### DISCUSSION THREAD

**Discussion by Prasunamadasu** [cryostatio/cryostat#739] [Q&A]
I have noticed that when a target becomes unavailable or its readiness probe fails, it still appears in the Cryostat target list. Can someone explain why this happens? Is there a specific reason Cryostat maintains visibility of these lost targets, and how does this affect its ability to monitor and interact with them?

Thank you!

**Comment by andrewazores** [comment]
It isn't supposed to do that, this is a bug. Which discovery mechanism are these targets found by? I guess since you mention readiness probes, it's Kubernetes discovery?

**Reply by andrewazores** [reply]
https://github.com/cryostatio/cryostat/pull/689

**Reply by andrewazores** [reply]
Also, I should mention that Kubernetes discovery works by looking for Endpoints objects in the discovery namespaces. These are basically just the Pod IP/Port pairs corresponding to Services in the namespaces, where the port name/number matches that Cryostat is looking for.

I'm not sure off the top of my head here what Kubernetes does with the Endpoints if the Pod is failing its readiness probe. If the Endpoints object still exists then Cryostat should still see the target. This part isn't a bug - the target could be failing its HTTP readiness probe, yet its JVM could actually be up and running and ready to accept Cryostat's JMX connections. And in fact, this might specifically be helpful if you want to use a JFR profiling session as part of diagnosing why the target is failing its readiness probe.

**Reply by Prasunamadasu** [reply]
Thank you for your detailed response. Yes, we are using Kubernetes discovery for these targets and using Java Management Extensions (JMX) to communicate with target applications.
Your explanation about Kubernetes Endpoints objects and how Cryostat uses them for discovery is very helpful. 

**Regarding our specific situation:**

1. We've observed that when a pod is restarted or deleted, its corresponding target sometimes remains visible in Cryostat for a while.
2. This has led to some confusion when trying to diagnose issues, as we weren't sure if the target was actually available or not. However, I can see how this behavior could be beneficial for debugging purposes, especially in cases where the JVM is running but failing HTTP readiness probes.
3. Additionally, we are facing an issue where creating recordings on certain targets results in a 504 Gateway Timeout error. This happens even though the target appears in Cryostat's list, which makes it difficult to determine if the problem is with Cryostat or the target application itself.

**A few questions:**

1. Is there a way to distinguish in Cryostat's UI between targets that are fully healthy and those that might be in this "failing readiness probe but JVM accessible" state?
2. Do you have any recommendations for troubleshooting the 504 Gateway Timeout issue when creating recordings?

**Reply by andrewazores** [reply]
About your situation:
1. This sounds just like #634 / #689, so it should be addressed in the latest `main` builds, or when those changes land in a release.

To answer the questions:
1. Cryostat doesn't know anything about readiness probes at this time, so there isn't a direct way to check this. This sounds like a good idea and would fit well with https://github.com/cryostatio/cryostat/issues/71, whenever there's time to get around to that. The only sort-of-proxy to this information is the JVM ID property that Cryostat stores for each discovered target. This is a hash of various MXBean data that Cryostat uses to try to identify cases where multiple discovered targets actually represent one running JVM instance (ex. there are multiple Service objects pointing to it, or multiple enabled discovery plugins). Cryostat tries to determine the JVM ID once when it first discovers the target, or post-#689 it will try again for some time until this attempt succeeds. If a target exists but does not have the JVM ID property set then you can identify these in the Cryostat UI's Topology view because they will be hightlighted in yellow with a warning icon. However, once Cryostat does determine a JVM ID for a target, then it stops trying to check it, and unless you do some action like trying to start a recording then Cryostat won't try to connect to the target again to check on it - so if the application was initially connectable and Cryostat determined its JVM ID, and then the application became unhealthy and started failing its readiness probe, then you'll still be in a situation where Cryostat sees the target and thinks it's OK, whereas Kubernetes will tell you the application is unhealthy.
2. There are only a few ways that Cryostat should respond with a 504:
- https://github.com/cryostatio/cryostat/blob/a1410726bc892a6e1122de75b462d128d3429258/src/main/java/io/cryostat/ExceptionMappers.java#L101
- https://github.com/cryostatio/cryostat/blob/a1410726bc892a6e1122de75b462d128d3429258/src/main/java/io/cryostat/targets/TargetConnectionManager.java#L236
- https://github.com/cryostatio/cryostat/blob/a1410726bc892a6e1122de75b462d128d3429258/src/main/java/io/cryostat/targets/TargetConnectionManager.java#L245

Usually, a 504 status means exactly what you've noticed here - that Cryostat knows about a target which no longer exists or is no longer connectable, so when it tries to open a JMX connection using the target connection URL, it fails/times out. This situation should be somewhat improved when 4.0 is released with #689, since these "stale" targets should be cleaned up and removed more reliably. A 504 response can also happen if the target connection URL is connectable, but the port on the other end doesn't seem to be a JMX port - it could be a misconfiguration of the target Service object, or a proxy, or some other network issue in between Cryostat and the target.

**Reply by andrewazores** [reply]
It also just occurred to me that this readiness probe idea is very related to https://github.com/cryostatio/cryostat/issues/717 .

**Reply by andrewazores** [reply]
^ https://github.com/cryostatio/cryostat/pull/740

**Reply by Prasunamadasu** [reply]
Thank you for your detailed response.

- I've confirmed that the JMX port is correctly configured.
- I can access Cryostat via port forwarding and have successfully added the credentials.
- Cryostat is able to determine the JVM ID.
- When attempting to create a recording, I'm still encountering the 504 Gateway Timeout error.

Upon checking the Cryostat pod logs, I've noticed a recurring warning that seems to be related to this issue:

> WARN [io.cry.tar.TargetConnectionManager] (executor-thread-118) java.lang.IllegalAccessError: class jdk.jfr.internal.SecuritySupport$$Lambda$2313/0x000000010183f840 (in module jdk.jfr) cannot access class com.singularity.ee.agent.appagent.entrypoint.bciengine.FastMethodInterceptorDelegatorBoot (in unnamed module @0x6615435c) because module jdk.jfr does not read unnamed module @0x6615435c

**Reply by andrewazores** [reply]
Interesting. I haven't seen anything like that before, but it sounds like it makes sense as the root cause for your issue. Are you doing anything in particular with JPMS or JLink in your application?

https://stackoverflow.com/questions/57457317/unnamed-modules-cannot-be-read

https://community.appdynamics.com/t5/Java-Java-Agent-Installation-JVM-and-Controller-Installation/APM-Agent-won-t-load-in-WebLogic-14-1/td-p/53583

Can you try adding `--add-reads jdk.jfr=ALL-UNNAMED` to your application's JVM arguments?

**Reply by andrewazores** [reply]
Somewhat similar: https://github.com/cryostatio/cryostat/issues/682
